vividus-plugin-mobitru/src/main/java/org/vividus/mobitru/client/MobitruFacadeImpl.java

```diff
@@ -54,39 +66,104 @@ public MobitruFacadeImpl(MobitruClient mobitruClient)
     @Override
     public String takeDevice(DesiredCapabilities desiredCapabilities) throws MobitruOperationException
     {
-        Waiter waiter = new DurationBasedWaiter(new WaitMode(waitForDeviceTimeout, RETRY_TIMES));
-        Device requestedDevice = new Device();
-        requestedDevice.setDesiredCapabilities(desiredCapabilities.asMap());
-        String capabilities = performMapperOperation(mapper -> mapper.writeValueAsString(requestedDevice));
-        byte[] receivedDevice = waiter.wait(() -> {
+        if (isSearchForDevice(desiredCapabilities))
+        {
+            DeviceSearchParameters deviceSearchParameters = new DeviceSearchParameters(desiredCapabilities);
+            List<Device> foundDevices = findDevices(deviceSearchParameters);
+            return takeDevice(foundDevices);
+        }
+        else
+        {
+            Device device = new Device();
+            device.setDesiredCapabilities(desiredCapabilities.asMap());
+            Waiter deviceWaiter = new DurationBasedWaiter(new WaitMode(waitForDeviceTimeout, RETRY_TIMES));
+            return takeDevice(device, deviceWaiter);
+        }
+    }
+
+    @Override
+    public void installApp(String deviceId, String appRealName) throws MobitruOperationException
+    {
+        mobitruClient.installApp(deviceId, findApp(appRealName));
+    }
+
+    @Override
+    public void returnDevice(String deviceId) throws MobitruOperationException
+    {
+        mobitruClient.returnDevice(deviceId);
+    }
+
+    private String takeDevice(Device device, Waiter deviceWaiter) throws MobitruOperationException
+    {
+        LOGGER.info("Trying to take {}", device);
+        String capabilities = performMapperOperation(mapper -> mapper.writeValueAsString(device));
+        byte[] receivedDevice = deviceWaiter.wait(() -> {
             try
             {
                 return mobitruClient.takeDevice(capabilities);
             }
-            catch (MobitruDeviceSearchException e)
+            catch (MobitruDeviceTakeException e)
             {
                 LOGGER.warn("Unable to take device, retrying attempt.", e);
                 return null;
             }
         }, Objects::nonNull);
         if (null == receivedDevice)
         {
-            throw new MobitruOperationException(String.format("Unable to find device using %s", desiredCapabilities));
+            throw new MobitruDeviceTakeException(String.format("Unable to take %s", device));
```

Comment:
Maybe in case of inability to get a device we need to be more collaborative and explain to user what had been done, like: unable to find device with these capabilities: ( these search parameters ) ?

Reply:
I think it's logged before the operation